### DISCUSSION TITLE
Does you Takeout have proper exifs?

### DISCUSSION THREAD

**Discussion by TheLastGimbus** [TheLastGimbus/GooglePhotosTakeoutHelper#191] [General]
I tried to take on inserting exifs again, and it, again, seems complicated/error-prone

**All** Takeouts that **I** took had exifs of original photos :thinking: 

- how are yours? Were they stripped? Or maybe becuse you uploaded photos sent through messenger (and potentially would want exif re-inserted)? 
- would you be satisifed to just run `exiftool "-AllDates<FileModifyDate" *` after gpth? it would be way safer, and could also work for videos!!

Please give me some feedback 🌈

**Comment by palijn** [comment]
OK, I was wrong and I sent you through an unneeded rabbit hole. My deepest apologies !

I finally got to use `exiftool -s -CreateDate "Photos from 2020/*` _(change folder date accordingly)_ to get the information right.
Result : either my takeout photos have **no** EXIF,
or they have EXIF createDate data that seems to match with the expected photo taken date .

I'll assume that when no EXIF is there, there should be some data in a corresponding JSON file.
A few random checks seem to show that when EXIF createDate data is already in the file, the phototakentime token in the json file matches. (I'm too lazy to program a thorough check of all such files…)

Next step will then be running your tool to sync the file modification date with the relevant creation date data.
Am I correct to assume that if there is a file with EXIF data and **no JSON** , then your tool will still set the file modification time to the createDate from EXIF ? Or, maybe such a case just never happens ?

Encouraged by your support, I finally got to dive into exiftool. 
I'm perfectly fine _now_ running `exiftool "-CreateDate< FileModifyDate" *` in every resulting folder afterwards.
(I want to fill EXIF data in all photos, since file modification time can vary sooo much with the smallest of interventions…)

**Reply by palijn** [reply]
Ah, yes, you're very right : it seems that photos lacking EXIF data have most probably been received through some messenger app which stripped the info. Dumb me never thought of that. Thanks again for the enlightenment !

**Reply by TheLastGimbus** [reply]
hey man no problem, very glad everything is fine :smile:

**Reply by TheLastGimbus** [reply]
> Am I correct to assume that if there is a file with EXIF data and no JSON , then your tool will still set the file modification time to the createDate from EXIF ?

yes, gpth tries to find correct date at all cost - it even guesses them from filenames if there is no exif nor json found :O

**Reply by palijn** [reply]
Update : I found out that vanilla exiftool creates a _original for each modified file, effectively doubling the storage space. Plus, the newly formed file, with proper EXIF, then has the modification time at the run of exiftool, therefore breaking all the nice work of gpth.
The proper command line is : `exiftool -overwrite_original -P "-CreateDate< FileModifyDate" *`

**Reply by palijn** [reply]
For the sake of full documentation : exiftool discovered an insane amount of issues with my Takeout files, most minor.

One of the errors was unexpected : some .jpg files were actually PNG contents, and some .png files were actually JPEG contents. Exiftool plain refused to operate on them, so, I had to rename them.

Script ran at the root folder of gpth-output : 

```
#/bin/bash
find . -type d | while read DIR ;
do 
echo "DIR is $DIR"

for f in "$DIR"/*.jpg ; do
  if [[ $(file -b --mime-type "$f") = image/png ]] ; then
   mv   "$f" "${f/%.jpg/.png}"
   echo "$f" "${f/%.jpg/.png}" 
  fi
done

for j in "$DIR"/*.png ; do
  if [[ $(file -b --mime-type "$j") = image/jpeg ]] ; then
   mv   "$j" "${j/%.png/.jpg}"
   echo "$j" "${j/%.png/.jpg}"
  fi

done

done
```

Either this can help someone in the same situation, or at least it can warn you of the kind of issues you might encounter if you ever decide to actually wrangle with the file contents !

**Reply by TheLastGimbus** [reply]
tried exiftool and seen this too, tho i'm not sure it's always right - sometimes it said that it thing something is a gif when it acutally **was** a png

**Reply by TheLastGimbus** [reply]
hey @palijn, i did some fixes here and there, and want to release them into new version 

i want to call it a day and mark this version as bug-free - could you do some testing for me? especially with windoza if you have it

here is the latest nightly: https://github.com/TheLastGimbus/GooglePhotosTakeoutHelper/actions/runs/4589368020

**Reply by palijn** [reply]
hey @TheLastGimbus , I downloaded v3.4.1 , does it contain the fixes you want me to test ?
I ran `gpth-macos --input Takeout --output Takeout-GPTH --albums duplicate-copy --divide-to-dates` on my (fresh, original, extracted for this test) Takeout folder and the `*-modifié*` still end up in the `date-unknown` folder.

`ll ALL_PHOTOS/date-unknown 
total 28328
-rwxr-xr-x@ 1 palijn  staff  268898  2 avr 22:56 IMG-20150125-WA0003-modifié.jpg*
`
`ll ALL_PHOTOS/2015/01/IMG-20150125-WA0003.jpg 
-rwxr-xr-x@ 1 palijn  staff  236264 25 jan  2015 ALL_PHOTOS/2015/01/IMG-20150125-WA0003.jpg*
`

So, the original was properly stored, the -modifié was not - while, obviously, there is quite a hint of a date in the filename itself !
Did I do something wrong ?

**Reply by TheLastGimbus** [reply]
Weird, it should match, i tested against that...

Can you please show me screenshot/details/something of which jsons should -modifie files correspond to? Their names etc

**Reply by palijn** [reply]
Mmh, I don't know what to show. There does not seem to be any json associated with the -modifié files - at least none I can find in the original Takeout tree after running gpth.

**Reply by TheLastGimbus** [reply]
The `-modified` etc files usually go like:

```
image.jpg
image-modified.jpg
Image.jpg.json <- one for both
```

Can you see json without `-modified` ?

// If not, then we are good, there's nothing more gpth could do 👍
// ..or maybe we can get "guess date without hour" from the name as last resort 🧐

**Comment by palijn** [comment]
Yes, that's why I use the file command to double-check in my little script. 

⁣Télécharger TypeApp pour Android ​

Le 2 avr. 2023 à 17:08, à 17:08, "Mateusz Soszyński" ***@***.***> a écrit:

**Comment by palijn** [comment]
yes, there is a json for the un-modifié file.

⁣Télécharger TypeApp pour Android ​

Le 4 avr. 2023 à 10:54, à 10:54, "Mateusz Soszyński" ***@***.***> a écrit:

**Reply by TheLastGimbus** [reply]
Geat, then show me it please

**Reply by palijn** [reply]
```
{
  "title": "IMG-20141225-WA0000.jpg",
  "description": "",
  "imageViews": "9",
  "creationTime": {
    "timestamp": "1419548266",
    "formatted": "25 déc. 2014, 22:57:46 UTC"
  },
  "photoTakenTime": {
    "timestamp": "1419538566",
    "formatted": "25 déc. 2014, 20:16:06 UTC"
  },
  "geoData": {
    "latitude": 0.0,
    "longitude": 0.0,
    "altitude": 0.0,
    "latitudeSpan": 0.0,
    "longitudeSpan": 0.0
  },
  "geoDataExif": {
    "latitude": 0.0,
    "longitude": 0.0,
    "altitude": 0.0,
    "latitudeSpan": 0.0,
    "longitudeSpan": 0.0
  },
  "url": "https://lh3.googleusercontent.com/33ogjYER8H_dTMFm87i0F1totPj6QUh8oYUTD3XUimY2QP_SWcecLII0gHgFepourecmM_qCVvslOwytPngDs4fh_oP4-MnDfKKpwrY",
  "googlePhotosOrigin": {
    "mobileUpload": {
      "deviceFolder": {
        "localFolderName": "WhatsApp Images"
      },
      "deviceType": "ANDROID_PHONE"
    }
  },
  "photoLastModifiedTime": {
    "timestamp": "1649419852",
    "formatted": "8 avr. 2022, 12:10:52 UTC"
  }
}
```

**Reply by palijn** [reply]
And there they are :
```
find . -name IMG-20141225-WA0000\* -print
./Takeout-GPTH/ALL_PHOTOS/2014/12/IMG-20141225-WA0000.jpg
./Takeout-GPTH/ALL_PHOTOS/date-unknown/IMG-20141225-WA0000-modifié.jpg
./Takeout/Google Photos/Photos from 2014/IMG-20141225-WA0000.jpg.json
```

**Reply by palijn** [reply]
btw (writing here because  I can't find a way to reach you privately) I'v made a small donation on Ko-Fi. Hopefully it went through to you ?

**Reply by TheLastGimbus** [reply]
Yes it did, thanks very much ^_^ !  I will look into those names in some time...

**Reply by palijn** [reply]
Hello @TheLastGimbus , have you tried to catch the -modifié files yet ?
I'm going to have to use GPTH on a large Takeout (300GB, # of photos still unknown, I'm uncompressing the zips…), and there will definitely be some in there.
